File: Labs/Lab 5/Lab5.py
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
      self.val = val
      self.left = left
      self.right = right

def verify(root):
  return verify_helper(root, float('-inf'), float('inf'))

def verify_helper(curr_node, lower_bound, upper_bound):
  if curr_node is None:
      return True
  if not lower_bound < curr_node.val < upper_bound:
      return False
  return (verify_helper(curr_node.left, lower_bound, curr_node.val) and
          verify_helper(curr_node.right, curr_node.val, upper_bound))

# Example usage
if __name__ == "__main__":
  node3 = TreeNode(3)
  node4 = TreeNode(4)
  node5 = TreeNode(5)
  node6 = TreeNode(6)
  node2 = TreeNode(2, node3, node4)
  node8 = TreeNode(8, node5, node6)
  root = TreeNode(1, node2, node8)
  print(verify(root))

  nodes1 = TreeNode(1)
  nodes3 = TreeNode(3)
  nodes6 = TreeNode(6)
  nodes8 = TreeNode(8)
  nodes2 = TreeNode(2, nodes1, nodes3)
  nodes7 = TreeNode(7, nodes6, nodes8)
  roots = TreeNode(5, nodes2, nodes7)
  print(verify(roots))

File: Labs/Lab 5/Lab5.java
public class Lab5 {
	
	public static boolean verify(TreeNode root) {
		return verifyHelper(root, Integer.MIN_VALUE , Integer.MAX_VALUE);
	}

	private static boolean verifyHelper(TreeNode currNode, int lowerBound, int upperBound) {
		if (currNode == null) {
			return true;
		}
		if (currNode.val <= lowerBound || currNode.val >= upperBound) {
			return false;
		}
		if (!verifyHelper(currNode.left, lowerBound, currNode.val)) {
			return false;
		}
		if (!verifyHelper(currNode.right, currNode.val, upperBound)) {
			return false;
		}
		return true;
	}

	public static void main(String[] args) {
		TreeNode node3 = new TreeNode(3);
		TreeNode node4 = new TreeNode(4);
		TreeNode node5 = new TreeNode(5);
		TreeNode node6 = new TreeNode(6);
		TreeNode node2 = new TreeNode(2, node3, node4);
		TreeNode node8 = new TreeNode(8, node5, node6);
		TreeNode root = new TreeNode(1, node2, node8);
		System.out.println(verify(root));

		TreeNode nodes1 = new TreeNode(1);
		TreeNode nodes3 = new TreeNode(3);
		TreeNode nodes6 = new TreeNode(6);
		TreeNode nodes8 = new TreeNode(8);
		TreeNode nodes2 = new TreeNode(2, nodes1, nodes3);
		TreeNode nodes7 = new TreeNode(7, nodes6, nodes8);
		TreeNode roots = new TreeNode(5, nodes2, nodes7);
		System.out.println(verify(roots));
	}
}


